ic-certification/src/hash_tree/mod.rs

```diff
@@ -123,100 +182,86 @@ impl<'a> HashTree<'a> {
     where
-        P: IntoIterator<Item = &'p Label>,
+        P: IntoIterator<Item = &'p I>,
+        I: ?Sized + AsRef<[u8]> + 'p,
     {
-        self.root.lookup_subtree(&mut path.into_iter())
+        self.root
+            .lookup_subtree(&mut path.into_iter().map(|v| v.borrow()))
     }
 
     /// List all paths in the [HashTree]
-    pub fn list_paths(&self) -> Vec<Vec<Label>> {
+    pub fn list_paths(&self) -> Vec<Vec<Label<Storage>>> {
         self.root.list_paths(&vec![])
     }
 }
 
-impl<'a> AsRef<HashTreeNode<'a>> for HashTree<'a> {
-    fn as_ref(&self) -> &HashTreeNode<'a> {
+impl<Storage: AsRef<[u8]>> AsRef<HashTreeNode<Storage>> for HashTree<Storage> {
+    fn as_ref(&self) -> &HashTreeNode<Storage> {
         &self.root
     }
 }
 
-impl<'a> From<HashTree<'a>> for HashTreeNode<'a> {
-    fn from(tree: HashTree<'a>) -> HashTreeNode<'a> {
+impl<Storage: AsRef<[u8]>> From<HashTree<Storage>> for HashTreeNode<Storage> {
+    fn from(tree: HashTree<Storage>) -> HashTreeNode<Storage> {
         tree.root
     }
 }
 
-#[cfg(feature = "serde")]
-impl serde::Serialize for HashTree<'_> {
-    fn serialize<S>(
-        &self,
-        serializer: S,
-    ) -> Result<<S as serde::Serializer>::Ok, <S as serde::Serializer>::Error>
-    where
-        S: serde::Serializer,
-    {
-        self.root.serialize(serializer)
-    }
-}
-
-#[cfg(feature = "serde")]
-impl<'de> serde::Deserialize<'de> for HashTree<'_> {
-    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
-    where
-        D: serde::Deserializer<'de>,
-    {
-        Ok(HashTree {
-            root: HashTreeNode::deserialize(deserializer)?,
-        })
-    }
-}
-
 /// Create an empty hash tree.
 #[inline]
-pub fn empty() -> HashTree<'static> {
+pub fn empty<Storage: AsRef<[u8]>>() -> HashTree<Storage> {
     HashTree {
         root: HashTreeNode::Empty(),
     }
 }
 
 /// Create a forked tree from two trees or node.
 #[inline]
-pub fn fork<'a, 'l: 'a, 'r: 'a>(left: HashTree<'l>, right: HashTree<'r>) -> HashTree<'a> {
+pub fn fork<Storage: AsRef<[u8]>>(
+    left: HashTree<Storage>,
+    right: HashTree<Storage>,
+) -> HashTree<Storage> {
     HashTree {
         root: HashTreeNode::Fork(Box::new((left.root, right.root))),
     }
 }
 
 /// Create a labeled hash tree.
 #[inline]
-pub fn label<'a, L: Into<Label>, N: Into<HashTree<'a>>>(label: L, node: N) -> HashTree<'a> {
+pub fn label<Storage: AsRef<[u8]>, L: Into<Label<Storage>>, N: Into<HashTree<Storage>>>(
+    label: L,
+    node: N,
+) -> HashTree<Storage> {
     HashTree {
-        root: HashTreeNode::Labeled(Cow::Owned(label.into()), Box::new(node.into().root)),
+        root: HashTreeNode::Labeled(label.into(), Box::new(node.into().root)),
     }
 }
 
 /// Create a leaf in the tree.
 #[inline]
-pub fn leaf<L: AsRef<[u8]>>(leaf: L) -> HashTree<'static> {
+pub fn leaf<Storage: AsRef<[u8]>, L: Into<Storage>>(leaf: L) -> HashTree<Storage> {
```

Comment:
These functions are probably going to be annoying in type inference. Not sure what the best solution is, but given as they're convenience functions, something sub-optimal is fine, like keeping them vec-only.

Reply:
That's why I made even more convenient `fn` "aliases" up in the top-level of the crate (i.e. `ic_certification::leaf`). Also, you can see the amount of typing you have to provide in `ic-certification/src/hash_tree/tests.rs`:

```rust
#[test]
fn works_with_simple_tree() {
    let tree: HashTree<Vec<u8>> = fork(
        label("label 1", empty()),
        fork(
            pruned(*b"\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01"),
            leaf([1u8, 2, 3, 4, 5, 6]),
        ),
    );

    assert_eq!(
        hex::encode(tree.digest()),
        "69cf325d0f20505b261821a7e77ff72fb9a8753a7964f0b587553bfb44e72532"
    );
}
```

Doesn't seem too bad to me.